fem/libceed/util.cpp

```diff
@@ -0,0 +1,346 @@
+   Mesh *mesh = fes.GetMesh();
+   const int ndofs = maps.ndof;
+   const int nqpts = maps.nqpt;
+   Vector qX(nqpts), qW(nqpts);
+   const IntegrationRule &ir1d = IntRules.Get(Geometry::SEGMENT, ir.GetOrder());
+   for (int i = 0; i < nqpts; i++)
+   {
+      const IntegrationPoint &ip = ir1d.IntPoint(i);
+      qX(i) = ip.x;
+      qW(i) = ip.weight;
+   }
+   CeedBasisCreateTensorH1(ceed, mesh->Dimension(), fes.GetVDim(), ndofs,
+                           nqpts, maps.Bt.GetData(),
+                           maps.Gt.GetData(), qX.GetData(),
+                           qW.GetData(), basis);
+}
+
+static void InitCeedTensorRestriction(const FiniteElementSpace &fes,
+                                      Ceed ceed, CeedElemRestriction *restr)
+{
+   Mesh *mesh = fes.GetMesh();
+   const FiniteElement *fe = fes.GetFE(0);
+   const TensorBasisElement * tfe =
+      dynamic_cast<const TensorBasisElement *>(fe);
+   MFEM_VERIFY(tfe, "invalid FE");
+   const Array<int>& dof_map = tfe->GetDofMap();
+
+   CeedInt compstride = fes.GetOrdering()==Ordering::byVDIM ? 1 : fes.GetNDofs();
+   const Table &el_dof = fes.GetElementToDofTable();
+   Array<int> tp_el_dof(el_dof.Size_of_connections());
+   const int dof = fe->GetDof();
+   const int stride = compstride == 1 ? fes.GetVDim() : 1;
+   if (dof_map.Size()>0)
+   {
+      for (int i = 0; i < mesh->GetNE(); i++)
+      {
+         const int el_offset = dof * i;
+         for (int j = 0; j < dof; j++)
+         {
+            tp_el_dof[j+el_offset] = stride*el_dof.GetJ()[dof_map[j]+el_offset];
+         }
+      }
+   }
+   else // dof_map.Size == 0, means dof_map[j]==j;
+   {
+      for (int i = 0; i < mesh->GetNE(); i++)
+      {
+         const int el_offset = dof * i;
+         for (int j = 0; j < dof; j++)
+         {
+            tp_el_dof[j+el_offset] = stride*el_dof.GetJ()[j+el_offset];
+         }
+      }
+   }
+   CeedElemRestrictionCreate(ceed, mesh->GetNE(), dof, fes.GetVDim(),
+                             compstride, (fes.GetVDim())*(fes.GetNDofs()),
+                             CEED_MEM_HOST, CEED_COPY_VALUES,
+                             tp_el_dof.GetData(), restr);
+}
+
+void InitCeedStridedRestriction(const FiniteElementSpace &fes,
+                                CeedInt nelem, CeedInt nqpts, CeedInt qdatasize,
+                                const CeedInt *strides,
+                                CeedElemRestriction *restr)
+{
+   CeedRestrKey restr_key(&fes, nelem, nqpts, qdatasize, restr_type::Strided);
+   auto restr_itr = internal::ceed_restr_map.find(restr_key);
+   if (restr_itr == internal::ceed_restr_map.end())
+   {
+      CeedElemRestrictionCreateStrided(internal::ceed, nelem, nqpts, qdatasize,
+                                       nelem*nqpts*qdatasize,
+                                       strides,
+                                       restr);
+      internal::ceed_restr_map[restr_key] = *restr;
```

Comment:
When are these objects freed? When the FESpace is destroyed? Is that the right choice?

Reply:
Due to the key for the map being associated to a `FiniteElementSpace` I thought so, the idea for the maps was to prevent duplicating potentially expensive `libCEED` objects. Do you have any proposition?

Reply:
I added a comment.